### ISSUE TITLE
Work thread not exit properly

### ISSUE THREAD

**Issue by dinglevin** [LMAX-Exchange/disruptor#194]
Checking below "***Test Code***"(let's leave the argument whether the code use Disruptor correctly or not later, focus on code first).

After "All exit", before sleep, if we use following command:
`jstack <pid> | grep "disruptor-consumer" `

We'll see the result like this, meaning these threads are still alive. 
```
"disruptor-consumer-480" prio=5 tid=0x00007fbdd886f000 nid=0x22203 waiting on condition [0x0000700003cca000]
"disruptor-consumer-478" prio=5 tid=0x00007fbdd88b1800 nid=0x22003 waiting on condition [0x0000700003ac4000]
"disruptor-consumer-477" prio=5 tid=0x00007fbdd88b2800 nid=0x21f03 waiting on condition [0x00007000038be000]
"disruptor-consumer-470" prio=5 tid=0x00007fbdda15a800 nid=0x21803 waiting on condition [0x00007000037bb000]
"disruptor-consumer-469" prio=5 tid=0x00007fbdda15a000 nid=0x21703 waiting on condition [0x00007000036b8000]
```

But if we halt disruptor multi times(consumer size) before next loop, all threads exit correctly.
```
        for (int i = 0; i < 100; i++) {
            Disruptor<String> disruptor = createDisruptor(32, 5);

            System.out.println("Wait to finish");

            disruptor.shutdown();
            for (int j = 0; j < 5; j++) {
                disruptor.halt();
            }

            System.out.println("Wait exit");
        }

        System.out.println("All exit");
```

If we check the problem thread call stack, the threads were waiting for BlockingWaitStrategy.processorNotifyCondition to be signaled, though, first disruptor halt(shutdown) signal all, but seems it is not always working as expected, that's why multiple halt seems works. One question left is why the first signal all not working? I am using Mac OS, the code also tested in Linux. What's more, if we sleep for several mins (not always working for 10s) before next loop, all threads can also exit correctly.
```
"disruptor-consumer-33" prio=5 tid=0x00007fddd4848000 nid=0x6003 waiting on condition [0x0000700001961000]
   java.lang.Thread.State: WAITING (parking)
	at sun.misc.Unsafe.park(Native Method)
	- parking to wait for  <0x00000007ee61d7f8> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
	at java.util.concurrent.locks.LockSupport.park(LockSupport.java:186)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2043)
	at com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
	at com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:56)
	at com.lmax.disruptor.WorkProcessor.run(WorkProcessor.java:148)
	at java.lang.Thread.run(Thread.java:722)
```

## Test Code:
```
public class MyDisruptorIssue {

    private static final AtomicInteger threadCounter = new AtomicInteger();
    private static final ThreadFactory threadFactory = new ThreadFactory() {

                                                         @Override
                                                         public Thread newThread(Runnable r) {
                                                             return new Thread(r, "disruptor-consumer-"
                                                                                  + threadCounter.incrementAndGet());
                                                         }
                                                     };

    public static void main(String[] args) throws Exception {
        for (int i = 0; i < 100; i++) {
            Disruptor<String> disruptor = createDisruptor(32, 5);

            System.out.println("Wait to finish");

            disruptor.shutdown();

            System.out.println("Wait exit");
        }

        System.out.println("All exit");

        Thread.sleep(100000000);
    }

    private static Disruptor<String> createDisruptor(int bufferSize, int consumerSize) {
        Disruptor<String> disruptor = new Disruptor<String>(getEventFactory(), bufferSize, threadFactory);

        WorkHandler<String>[] workHandlers = new WorkHandler[consumerSize];
        for (int i = 0; i < workHandlers.length; i++) {
            workHandlers[i] = new MyWorkHandler();
        }
        disruptor.handleEventsWithWorkerPool(workHandlers);

        disruptor.setDefaultExceptionHandler(new ExceptionHandler<String>() {

            @Override
            public void handleEventException(Throwable ex, long sequence, String event) {
                System.err.println("handle exception on event: " + event + ", sequence: " + sequence);
                ex.printStackTrace();
            }

            @Override
            public void handleOnStartException(Throwable ex) {
                System.err.println("handle exception on start");
                ex.printStackTrace();
            }

            @Override
            public void handleOnShutdownException(Throwable ex) {
                System.err.println("handle exception on shutdown");
                ex.printStackTrace();
            }
        });

        disruptor.start();

        return disruptor;
    }

    private static class MyWorkHandler<T> implements WorkHandler<T>, LifecycleAware {

        @Override
        public void onEvent(T event) throws Exception {
            System.out.println("Processing event: " + event);
        }

        @Override
        public void onStart() {
            System.out.println(Thread.currentThread().getName() + " - started");
        }

        @Override
        public void onShutdown() {
            System.out.println(Thread.currentThread().getName() + " - shutdown");
        }
    }

    private static EventFactory<String> getEventFactory() {
        return new EventFactory<String>() {

            @Override
            public String newInstance() {
                return new String("");
            }
        };
    }
}
```

**Comment by mikeb01** [comment]
This is duplicate of #166.  Disruptor.shutdown is being called before the threads have had a chance to start.  You can use a combination of a latch and the LifecycleAware interface.  There is a link to an example in the aforementioned issue.

**Comment by dinglevin** [comment]
Ah, yes, thanks
